x-pack/filebeat/input/azureeventhub/connection_string.go

```diff
@@ -0,0 +1,133 @@
+func parseConnectionString(connStr string) (ConnectionStringProperties, error) {
+	const (
+		endpointKey              = "Endpoint"
+		sharedAccessKeyNameKey   = "SharedAccessKeyName"
+		sharedAccessKeyKey       = "SharedAccessKey"
+		entityPathKey            = "EntityPath"
+		sharedAccessSignatureKey = "SharedAccessSignature"
+		useEmulator              = "UseDevelopmentEmulator"
+	)
+
+	csp := ConnectionStringProperties{}
+
+	splits := strings.Split(connStr, ";")
+
+	for _, split := range splits {
+		if split == "" {
+			continue
+		}
+
+		keyAndValue := strings.SplitN(split, "=", 2)
+		if len(keyAndValue) < 2 {
+			return ConnectionStringProperties{}, errors.New("failed parsing connection string due to unmatched key value separated by '='")
+		}
+
+		// if a key value pair has `=` in the value, recombine them
+		key := keyAndValue[0]
+		value := strings.Join(keyAndValue[1:], "=")
+		switch {
+		case strings.EqualFold(endpointKey, key):
+			u, err := url.Parse(value)
+			if err != nil {
+				return ConnectionStringProperties{}, errors.New("failed parsing connection string due to an incorrectly formatted Endpoint value")
+			}
+			csp.Endpoint = value
+			csp.FullyQualifiedNamespace = u.Host
+		case strings.EqualFold(sharedAccessKeyNameKey, key):
+			csp.SharedAccessKeyName = &value
+		case strings.EqualFold(sharedAccessKeyKey, key):
+			csp.SharedAccessKey = &value
+		case strings.EqualFold(entityPathKey, key):
+			csp.EntityPath = &value
+		case strings.EqualFold(sharedAccessSignatureKey, key):
+			csp.SharedAccessSignature = &value
+		case strings.EqualFold(useEmulator, key):
+			v, err := strconv.ParseBool(value)
+
+			if err != nil {
+				return ConnectionStringProperties{}, err
+			}
+
+			csp.Emulator = v
+		}
+	}
+
+	if csp.Emulator {
+		// check that they're only connecting to localhost
+		endpointParts := strings.SplitN(csp.Endpoint, ":", 3) // allow for a port, if it exists.
+
+		if len(endpointParts) < 2 || endpointParts[0] != "sb" || endpointParts[1] != "//localhost" {
+			// there should always be at least two parts "sb:" and "//localhost"
+			// with an optional 3rd piece that's the port "1111".
+			// (we don't need to validate it's a valid host since it's been through url.Parse() above)
+			return ConnectionStringProperties{}, fmt.Errorf("UseDevelopmentEmulator=true can only be used with sb://localhost or sb://localhost:<port number>, not %s", csp.Endpoint)
+		}
+	}
+
+	if csp.FullyQualifiedNamespace == "" {
+		return ConnectionStringProperties{}, fmt.Errorf("key %q must not be empty", endpointKey)
+	}
+
+	if csp.SharedAccessSignature == nil && csp.SharedAccessKeyName == nil {
+		return ConnectionStringProperties{}, fmt.Errorf("key %q must not be empty", sharedAccessKeyNameKey)
+	}
+
+	if csp.SharedAccessKey == nil && csp.SharedAccessSignature == nil {
+		return ConnectionStringProperties{}, fmt.Errorf("key %q or %q cannot both be empty", sharedAccessKeyKey, sharedAccessSignatureKey)
+	}
```

Comment:
[Suggestion]
Perform all checks and accumulate the errors to return them all at once. It's rather frustrating from the user perspective to fix a validation error just to discover there is still another to be fixed.

Reply:
Yeah good point on accumulating errors. But could be a follow up to improve the error message to reduce scope of the PR :)